tun: add ioctl to set permanent hardware address

Provide an interface for userspace to set the permanent hardware
address on tun interfaces.

Signed-off-by: "Charles (Chas) Williams" <ciwillia@vyatta.att-mail.com>

---
 drivers/net/tun.c           |   12 +++++++++++-
 include/uapi/linux/if_tun.h |    2 ++
 2 files changed, 13 insertions(+), 1 deletion(-)

--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -3081,7 +3081,7 @@ static long __tun_chr_ioctl(struct file
 	int ret;
 	bool do_notify = false;
 
-	if (cmd == TUNSETIFF || cmd == TUNSETQUEUE ||
+	if (cmd == TUNSETIFF || cmd == TUNSETQUEUE || cmd == TUNSETPERMADDR ||
 	    (_IOC_TYPE(cmd) == SOCK_IOC_TYPE && cmd != SIOCGSKNS)) {
 		if (copy_from_user(&ifr, argp, ifreq_len))
 			return -EFAULT;
@@ -3372,6 +3372,12 @@ static long __tun_chr_ioctl(struct file
 		ret = open_related_ns(&net->ns, get_net_ns);
 		break;
 
+	case TUNSETPERMADDR:
+		/* Set permanent hw address */
+		ether_addr_copy(tun->dev->perm_addr, ifr.ifr_hwaddr.sa_data);
+		ret = 0;
+		break;
+
 	default:
 		ret = -EINVAL;
 		break;
--- a/include/uapi/linux/if_tun.h
+++ b/include/uapi/linux/if_tun.h
@@ -62,6 +62,8 @@
 #define TUNSETCARRIER _IOW('T', 226, int)
 #define TUNGETDEVNETNS _IO('T', 227)
 
+#define TUNSETPERMADDR _IOW('T', 230, struct ifreq)
+
 /* TUNSETIFF ifr flags */
 #define IFF_TUN		0x0001
 #define IFF_TAP		0x0002
